src/lib.rs

```diff
@@ -99,22 +111,26 @@ pub trait Program: Send + 'static {
   ///              the `x`th pixel in the `y`th row.
   fn render(&mut self, _pixels: &mut [Pixel]) {}
 
-  /// Synthesize audio
+  /// The program's synthesizer
   ///
-  /// Called by the runtime as needed to fill the outgoing audio buffer
+  /// Will be called by the runtime during initialization. If it returns
+  /// Some, the contained Synthesizer will be moved to a dedicated audio
+  /// thread and called periodically to produce samples for the outgoing
+  /// audio stream.
   ///
-  /// * `played`  — number of samples written by previous calls to synthesize
-  /// * `buffer`  — an array of audio samples
-  fn synthesize(&mut self, _samples_played: u64, _buffer: &mut [Sample]) {}
+  /// In order to prevent buffer underruns, avoid locking the `Mutex`
+  /// containing the Synthesizer for long periods of time.
+  fn synthesizer(&self) -> Option<Arc<Mutex<Synthesizer>>> {
```

Comment:
Would it be possible to have something like `fn synthesizer(&self) -> Option<Synthesizer> { None }`? And require that `Synthesizer` implements `Clone`? And then the library would call `sythesizer` after every call to `render`, clone it and update a mutex that it maintains itself? That way the API for users might be simpler.